File: TotalCrossVM/src/nm/phone/CellInfo.c
// Copyright (C) 2000-2013 SuperWaba Ltda.
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only



#include "tcvm.h"

#if defined WP8

#elif defined WINCE || defined WIN32
 #include "win/CellInfo_c.h"
#elif defined(ANDROID)
 #include "android/CellInfo_c.h"
#else
 //#include "posix/CellInfo_c.h"
#endif

// static fields
static TCClass cellInfoClass;
#define ConnectionManager_connRef(c)      getStaticFieldObject(null,c, "connRef")

#define CellInfo_cellId(c)                *getStaticFieldObject(null,c, "cellId")
#define CellInfo_mnc(c)                   *getStaticFieldObject(null,c, "mnc")
#define CellInfo_mcc(c)                   *getStaticFieldObject(null,c, "mcc")
#define CellInfo_lac(c)                   *getStaticFieldObject(null,c, "lac")
#define CellInfo_signal(c)                *getStaticFieldInt(c, "signal")

//////////////////////////////////////////////////////////////////////////
TC_API void tpCI_loadResources(NMParams p) // totalcross/phone/CellInfo native private void loadResources();
{
   cellInfoClass = loadClass(p->currentContext, "totalcross.phone.CellInfo", false);
#if defined (WINCE)
   CellInfoLoadResources(p->currentContext);
#endif
}
//////////////////////////////////////////////////////////////////////////
TC_API void tpCI_releaseResources(NMParams p) // totalcross/phone/CellInfo native private void releaseResources();
{
#if defined (WINCE)
  CellInfoReleaseResources();
#endif
}
//////////////////////////////////////////////////////////////////////////
TC_API void tpCI_update(NMParams p) // totalcross/phone/CellInfo native public static void update();
{                                        
#if defined (WINCE) || defined (ANDROID)
   int32 mcc=0, mnc=0, lac=0, cellId=0, signal=0;
   IntBuf ibuf;
   
   CellInfoUpdate(&mcc, &mnc, &lac, &cellId, &signal);
   CellInfo_mcc(cellInfoClass) = mcc == 0 ? null : createStringObjectFromCharP(p->currentContext, int2str(mcc, ibuf), -1);
   CellInfo_mnc(cellInfoClass) = mnc == 0 ? null : createStringObjectFromCharP(p->currentContext, int2str(mnc, ibuf), -1);
   CellInfo_lac(cellInfoClass) = lac == 0 ? null : createStringObjectFromCharP(p->currentContext, int2str(lac, ibuf), -1);
   CellInfo_cellId(cellInfoClass) = cellId == 0 ? null : createStringObjectFromCharP(p->currentContext, int2str(cellId, ibuf), -1);
   CellInfo_signal(cellInfoClass) = signal;       
#endif
}

#ifndef ANDROID
//////////////////////////////////////////////////////////////////////////
TC_API void tmA_getHeightD_i(NMParams p) // totalcross/money/Ads native static int getHeightD(int size);
{
   p->retI = 0;
}
//////////////////////////////////////////////////////////////////////////
TC_API void tmA_configureD_s(NMParams p) // totalcross/money/Ads native static void configureD(String id);
{
}
//////////////////////////////////////////////////////////////////////////
TC_API void tmA_setSizeD_i(NMParams p) // totalcross/money/Ads native static void setSizeD(int s);
{
}
//////////////////////////////////////////////////////////////////////////
TC_API void tmA_setPositionD_i(NMParams p) // totalcross/money/Ads native static void setPositionD(int p);
{
}
//////////////////////////////////////////////////////////////////////////
TC_API void tmA_setVisibleD_b(NMParams p) // totalcross/money/Ads native static void setVisibleD(boolean b);
{
}
//////////////////////////////////////////////////////////////////////////
TC_API void tmA_isVisibleD(NMParams p) // totalcross/money/Ads native static boolean isVisibleD();
{
   p->retI = 0;
}
#endif

File: TotalCrossSDK/src/main/java/totalcross/phone/CellInfo.java
// Copyright (C) 2000-2013 SuperWaba Ltda.
// Copyright (C) 2014-2020 TotalCross Global Mobile Platform Ltda.
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only

package totalcross.phone;

import com.totalcross.annotations.ReplacedByNativeOnDeploy;
import totalcross.io.ByteArrayStream;
import totalcross.io.DataStream;
import totalcross.io.LineReader;
import totalcross.sys.Convert;
import totalcross.sys.Settings;
import totalcross.sys.Vm;

/** Contains information about the anthena that this cell phone is receiving signal. 
 * Used by the GPS class.
 * @since TotalCross 1.22
 */

public class CellInfo {
  public static String cellId;
  public static String mnc;
  public static String mcc;
  public static String lac;
  public static int signal;

  static CellInfo instance = new CellInfo();

  private CellInfo() {
    loadResources();
  }

  @ReplacedByNativeOnDeploy
  public static void update() {
  }

  @ReplacedByNativeOnDeploy
  private void loadResources() {
  }

  @ReplacedByNativeOnDeploy
  private void releaseResources() {
  }

  @Override
  protected void finalize() {
    releaseResources();
  }

  public static boolean isSupported() {
    return Settings.isWindowsCE() || Settings.platform.equals(Settings.ANDROID);
  }

  /** Converts the current cellId and lac into the latitude and longitude.
   * You need to manually update the values by calling the <code>update</code> method.
   * 
   * This method requires direct connection to the internet and uses the google mmap service.
   * @since TotalCross 1.3
   */
  public static double[] toCoordinates() throws Exception {
    double[] ret = null;

    if (cellId == null || lac == null) {
      return null;
    }

    ByteArrayStream bas = new ByteArrayStream(256);
    DataStream ds = new DataStream(bas);

    ds.writeBytes("POST /glm/mmap HTTP/1.1\r\n");
    ds.writeBytes("Content-Type: application/binary\r\n");
    ds.writeBytes("Content-Length: 72\r\n");
    ds.writeBytes("\r\n");
    ds.writeShort(21);
    ds.writeLong(0);
    ds.writeShort(2);
    ds.writeBytes("en");
    ds.writeShort(7);
    ds.writeBytes("Android");
    ds.writeShort(3);
    ds.writeBytes("1.0");
    ds.writeShort(3);
    ds.writeBytes("Web");
    ds.writeByte(27);
    ds.writeInt(0);
    ds.writeInt(0);
    ds.writeInt(3);
    ds.writeShort(0);
    ds.writeInt(Convert.toInt(cellId));
    ds.writeInt(Convert.toInt(lac));
    ds.writeInt(0);
    ds.writeInt(0);
    ds.writeInt(0);
    ds.writeInt(0);

    totalcross.net.Socket sock = new totalcross.net.Socket("www.google.com", 80, 20000);
    sock.writeBytes(bas.toByteArray());

    Vm.sleep(250); // wait for a response

    // read the whole answer into a byte array
    byte[] buf = new byte[1024];
    int total = sock.readBytes(buf);
    bas = new ByteArrayStream(buf, total);
    ds = new DataStream(bas);

    LineReader lr = new LineReader(bas);
    lr.returnEmptyLines = true;
    String line = lr.readLine(); // first must be HTTP/1.1 200 OK
    if (line != null && line.indexOf(" 200 ") >= 0) {
      // find out the length of the data-chunk
      int len = 0;
      while ((line = lr.readLine()) != null) // skip all other header information
      {
        if (line.toLowerCase().startsWith("content-length:")) {
          len = Convert.toInt(line.substring(16).trim());
          break;
        } else if (line.equals("")) {
          break;
        }
      }
      if (len > 0) {
        // point to the data
        bas.setPos(total - len);
        ds.readShort();
        ds.readByte();
        int code = ds.readInt();
        if (code == 0) {
          ret = new double[] { ds.readInt() / 1e6d, ds.readInt() / 1e6d };
        }
      }
    }

    sock.close();

    return ret;
  }
}
